File: elixir/abc363/c/lib/main.ex
defmodule Main do
  def main do
    [n, k] = read_ints()
    s = IO.read(:line) |> String.trim()

    ans =
      s
      |> String.graphemes()
      |> Enum.sort()
      |> distinct_permutations()
      |> Stream.map(&Enum.join(&1))
      |> Stream.map(fn t -> 0..(n - k) |> Stream.map(&String.slice(t, &1, k)) end)
      |> Stream.map(fn l -> l |> Stream.map(&(&1 != String.reverse(&1))) end)
      |> Stream.map(&Enum.all?(&1))
      |> Enum.count(& &1)

    IO.puts(ans)
  end

  def read_ints() do
    IO.read(:line)
    |> String.trim()
    |> String.split(" ")
    |> Enum.map(&String.to_integer/1)
  end

  # Algorithm: https://w.wiki/Qai
  def distinct_permutations(s) do
    step1 = fn t ->
      t
      |> Enum.chunk_every(2, 1, :discard)
      |> Enum.find(fn [{b, _}, {a, _}] -> a < b end)
      |> then(fn c ->
        case c do
          [{_, _}, {_, _} = a] -> a
          _ -> throw(:exit)
        end
      end)
    end

    step2 = fn t, a ->
      t
      |> Enum.find(fn {b, _} -> a < b end)
      |> then(fn c ->
        case c do
          {_, _} = b -> b
          _ -> throw(:exit)
        end
      end)
    end

    step3 = fn s, {a, k}, {b, l} ->
      s
      |> List.replace_at(k, b)
      |> List.replace_at(l, a)
    end

    step4 = fn u, k ->
      (u |> Enum.slice(0..k)) ++ (u |> Enum.slice((k + 1)..-1//1) |> Enum.reverse())
    end

    Stream.iterate({:cont, s}, fn {_, s} ->
      try do
        t = s |> Enum.with_index() |> Enum.reverse()
        {a, k} = step1.(t)
        {b, l} = step2.(t, a)
        u = s |> step3.({a, k}, {b, l}) |> step4.(k)
        {:cont, u}
      catch
        :exit -> {:halt, []}
      end
    end)
    |> Stream.take_while(fn {c, _} -> c == :cont end)
    |> Stream.map(fn {_, s} -> s end)
  end
end

File: go/abc363/c/main.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	var n, k int
	var s []byte
	fmt.Scan(&n, &k, &s)

	var ans int
	sort.Slice(s, func(i, j int) bool { return s[i] < s[j] })
	for {
		var f bool
		for i := 0; i <= n-k; i++ {
			sub := s[i : i+k]
			if isPalindrome(sub) {
				f = true
				break
			}
		}

		if !f {
			ans++
		}

		if !nextPermutation(s) {
			break
		}
	}

	fmt.Println(ans)
}

func nextPermutation(s []byte) bool {
	step1 := func(s []byte) int {
		k := -1
		for i := len(s) - 2; i >= 0; i-- {
			if s[i] < s[i+1] {
				k = i
				break
			}
		}
		return k
	}

	step2 := func(s []byte, k int) int {
		l := -1
		for i := len(s) - 1; i > k; i-- {
			if s[k] < s[i] {
				l = i
				break
			}
		}
		return l
	}

	step3 := func(s []byte, k, l int) {
		s[k], s[l] = s[l], s[k]
	}

	step4 := func(s []byte, k int) {
		for i, j := k+1, len(s)-1; i < j; i, j = i+1, j-1 {
			s[i], s[j] = s[j], s[i]
		}
	}

	k := step1(s)
	if k < 0 {
		return false
	}
	l := step2(s, k)
	step3(s, k, l)
	step4(s, k)

	return true
}

func isPalindrome(s []byte) bool {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		if s[i] != s[j] {
			return false
		}
	}
	return true
}
